protocol/rep/rep.go

```diff
@@ -227,9 +225,8 @@ func (r *rep) RecvHook(m *mangos.Message) bool {
 	}
 	r.sock.SetSendError(nil)
 	r.backtraceL.Lock()
-	r.backtrace = append(r.backtracebuf[0:0], m.Header...)
+	r.backtrace = append(r.backtrace[0:0], m.Header...)
 	r.backtraceL.Unlock()
-	m.Header = nil
```

Comment:
You really need a unique request ID. If it were me I'd just add a 128 bit UUID to the top of every request.  Maybe we should look at adding something like this to the protocol. 

Sent from my iPhone

> On May 19, 2016, at 6:54 PM, Gerrit Renker <notifications@github.com> wrote:
> 
> In protocol/rep/rep.go:
> 
> >  	r.backtraceL.Unlock()
> > -	m.Header = nil
> I re-read my original commit message and admittedly, it is not very clear; I hope the following will be clearer.
> 
> Yes, on the req/rep, level the responses should be sent for each request, regardless of duplication.
> 
> The problem that I am attempting to solve is this. I am using the req/rep socket for communication with a worker node, where each request carries one command for the worker to execute.
> 
> The rationale was: when handing the request to the next-higher (application) level, there should be only 1 message passed up the layer, not 1 for each duplicate message. I am using a channel/callback to pass the received request upwards.
> 
> I had used the following approach, which I now realize was indeed naive:
> 
> use tuple of (first 4 bytes of msg.Header (endpoint channel ID) and last 4 bytes (request ID)) to index map[]bool
> before passing up the request, check if it has been seen before
> The problem with the above is that it will fail if there is a reconnect in between.
> 
> Still, I would like to avoid re-implementing unique request IDs. If the requesting endpoint has not rebooted, retransmits are "reasonably unique" (if avoiding sending so many messages that the 4-byte counter wraps around).
> 
> There is separate work on making the commands idempotent. I believe it would help with the problem I am working on, and it may be of similar help to others.
> 
> The spec does not make this mandatory, it just says
> 
> After getting hold on the request, the REP socket should prepend it  by 32 bit value, 
> consisting of 1 bit set to 0 followed by the 31-bit  ID of the channel the request was received from.  
> The extended  request will be then handed to the user.
> Would it be possible to let the application inspect the header, and what would be the format you would think of (I am thinking in terms of inspecting TCP/UDP connections using conn.RemoteAddr())?
> 
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
>